[FLINK-29870] [ResourceManager]Split ResourceActions to ResourceAllocator and ResourceEventListener.

huwh opened the pull request:
## What is the purpose of the change

SlotManager determines whether a resource can be allocated by the return value of ResourceActions.allocateResource. It can reduce unnecessary waiting by quickly marking the slotRequest as a failure if the resource cannot be allocated.

The new declareResourceNeed function will not return whether resources could be allocated. This will cause useless waiting for SlotRequests.

This PR will split ResourceNotEnoughNotifier from ResourceActions so that we can make ResourceActions only co-works with ActiveResourceManager. And SlotManager cloud determine whether can allocate new resources by ResourceActions is null or not.


## Brief change log
- Extract the notifyNotEnoughResourcesAvailable function from ResourceActions to ResourceNotEnoughNotifier
- Move ResourceActionsImpl from ResourceManager to ActiveResourceManager
- SlotManager determines whether a resource can be allocated by whether the ResourceActions is null or not.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes, ResourceManager)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
    * part of https://docs.google.com/document/d/1lcmf3MKmcmf9tsPc1whaZHMYKurGoGtqOXEep9ngP2k/edit#

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java:
```diff
@@ -86,7 +88,7 @@ public interface SlotManager extends AutoCloseable {
             ResourceManagerId newResourceManagerId,
             Executor newMainThreadExecutor,
-            ResourceActions newResourceActions,
+            @Nullable ResourceActions newResourceActions,
```
I would not make this argument `@Nullable`.

I'm aware that many fields, including `resourceActions`, are `@Nullable` in the slot manager implementations. However, that is because these fields can only be decided when the slot manager is started, not when constructed. The null values can be ambiguous. E.g., in this case, `resourceActions` being `null` can mean either the slot manager is not yet started, or the resource allocation is not supported.

As an alternative, I'd suggest to introduce another implementation of `ResourceActions`, something like a `NonSupportedResourceActions`. And we can add a new method `ResourceActions#isSupported`, and use it for deciding whether the resource can be allocated. For the `NonSupportedResourceActions`, `allocateResource` and `releaseResource` should throw `UnsupportedOperationException`, meaning that the slot manager is responsible for calling these two methods only on when `ResourceActions#isSupported` returns `true`.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManager.java:
```diff
@@ -499,6 +506,21 @@ public void execute(Runnable command) {
+        @Override
+        public boolean allocateResource(WorkerResourceSpec workerResourceSpec) {
+            validateRunsInMainThread();
+            return startNewWorker(workerResourceSpec);
```
`startNewWorker` is no longer needed for `ResourceManager` and `StandaloneResourceManager`.

xintongsong replied:
We probably have the same problem for `stopWorker`. There are some logics in `ResourceManager#releaseResource` that are not only for the active resource manager, however, the method `releaseResource` itself is never called except in the active resource manager.

xintongsong reviewed (changes requested):
Thanks for opening this PR, @huwh. I have left some comments, please take a look.

huwh commented:
> 
Thanks for reviewing, @xintongsong, i have updated this pr according to your suggestions

1. split a new issues to remove unused ResourceActions#notifyAllocationFailure in https://issues.apache.org/jira/browse/FLINK-30084. we need merge this PR at first, then i will rebase my changes in this pr.
3. rename the ResourceActions and ResourceNotEnoughNotifier
4. introduce new implementation of ResourceAllocator: NonSupportedResourceAllocatorImpl
5. move the ResourceManager#startNewWorker, releaseResource to ActiveResourceManager
    4.1 ResourceManager#releaseResource For non-ActiveResourceManager logic is not used, we can assume that TaskMangers registered on SlotManager must have been registered on ResourceManager, and when TaskManager is unregistered from RM, it will also be unregistered from SlotManager

xintongsong commented:
Well, I meant removing `notifyAllocationFailure` in a separate hotfix commit in this PR, to keep the commit history clean. Never mind. I Have merged the PRfor FLINK-30084.

Could you please rebase this PR and re-organize the commits. When addressing comments, it would be nice to keep the commits organized as to be merged. You can either address comments in fixup commits that clearly shows which previous commits they should be squashed into, or modify the previous commits and make a force-push if the comments cannot easily be addressed with fixup commits.

huwh commented:
> Well, I meant removing `notifyAllocationFailure` in a separate hotfix commit in this PR, to keep the commit history clean. Never mind. I Have merged the PRfor [FLINK-30084](https://issues.apache.org/jira/browse/FLINK-30084).
> 
> Could you please rebase this PR and re-organize the commits. When addressing comments, it would be nice to keep the commits organized as to be merged. You can either address comments in fixup commits that clearly shows which previous commits they should be squashed into, or modify the previous commits and make a force-push if the comments cannot easily be addressed with fixup commits.

Hi, @xintongsong, I've rebased this PR, but it's not easy to automatically squash with fixup commits because there are lots of classes rename. So, I modified the previous commit to addressing comments.
Could you please review this again in your free time. Thanks a lot.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManager.java:
```diff
@@ -299,6 +295,29 @@ public void onError(Throwable exception) {
+     * @return whether the resource can be allocated
+     */
+    @VisibleForTesting
+    public boolean startNewWorker(WorkerResourceSpec workerResourceSpec) {
+        requestNewWorker(workerResourceSpec);
+        return true;
+    }
```
It seems we no longer need this boolean return value.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -1262,16 +1252,6 @@ protected abstract void internalDeregisterApplication(
-     * @return whether the resource can be allocated
-     */
-    @VisibleForTesting
-    public abstract boolean startNewWorker(WorkerResourceSpec workerResourceSpec);
```
Shall we also move `stopWorker` to `ActiveResourceManager`? `StandaloneResourceManager#stopWorker` does nothing.

xintongsong replied:
And maybe `workerStarted` as well. The method is used for deciding whether to accept a TM registration. For active clusters we accept only the resources that we actively requested, while for standalone clusters we always accept.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/NonSupportedResourceAllocatorImpl.java:
```diff
@@ -0,0 +1,40 @@
+import org.apache.flink.runtime.resourcemanager.WorkerResourceSpec;
+
+/** ResourceAllocator that not support to allocate/release resources. */
+public class NonSupportedResourceAllocatorImpl implements ResourceAllocator {
```
This can be singleton.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/ResourceAllocator.java:
```diff
@@ -18,15 +18,13 @@
-public interface ResourceActions {
+public interface ResourceAllocator {
+
+    boolean isSupported();
```
JavaDoc is missing.

xintongsong reviewed (commented):
@huwh,
Thanks for addressing my comment. The PR looks much better. I have only a few more comments regarding the last commit. Please take a look.

Moreover, the CI failed due to code format issues. See [this page](https://nightlies.apache.org/flink/flink-docs-master/docs/flinkdev/ide_setup/#required-plugins) for setting IDEs to automatically format the codes.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java:
```diff
@@ -754,15 +754,16 @@ private boolean allocateResource(PendingTaskManager pendingTaskManager) {
-                        WorkerResourceSpec.fromTotalResourceProfile(
-                                pendingTaskManager.getTotalResourceProfile(),
-                                pendingTaskManager.getNumSlots()))) {
+        if (!resourceAllocator.isSupported()) {
```
It would be better to move this check to before the max total resource check, because this one is cheaper.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/NonSupportedResourceAllocatorImpl.java:
```diff
@@ -18,11 +18,14 @@
 /** ResourceAllocator that not support to allocate/release resources. */
 public class NonSupportedResourceAllocatorImpl implements ResourceAllocator {
+    public static final NonSupportedResourceAllocatorImpl INSTANCE =
+            new NonSupportedResourceAllocatorImpl();
```
For singleton classes, we should add a private constructor to prevent it from being instantiated.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java:
```diff
@@ -341,15 +348,17 @@ public boolean registerTaskManager(
-                        taskExecutorConnection.getInstanceID(),
-                        new FlinkExpectedException(
-                                "The max total resource limitation is reached."));
+                if (resourceAllocator.isSupported()) {
```
If- statements can be combined.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/TaskExecutorManager.java:
```diff
@@ -154,13 +154,15 @@ public boolean registerTaskManager(
-                    taskExecutorConnection.getInstanceID(),
-                    new FlinkExpectedException(
-                            "The total number of slots exceeds the max limitation."));
+            if (resourceAllocator.isSupported()) {
```
Same here. The if-statements can be combined.

xintongsong reviewed (approved):
Thanks @huwh. The PR LGTM in general. I'm taking over from here.

- I have squashed your previous fixup commits, and rebased the latest master branch.
- I edited the commit messages and added the component tag `[runtime]`.
- I also have some minor comments, which I addressed myself in another fixup commit. Please take a look.

If there's no further questions, I'll merge this once the CI passes.

huwh commented:
> Thanks @huwh. The PR LGTM in general. I'm taking over from here.
> 
> * I have squashed your previous fixup commits, and rebased the latest master branch.
> * I edited the commit messages and added the component tag `[runtime]`.
> * I also have some minor comments, which I addressed myself in another fixup commit. Please take a look.
> 
> If there's no further questions, I'll merge this once the CI passes.

@xintongsong Thanks a lot, I have no more questions, I will pay attention to these code style.
